README.md
# GCVideo #

GCVideo is a small series of FPGA boards and VHDL projects
capable of converting the
GameCube's Digital Video port signals to standard video signals
without using the custom chip in the original component video cable.
There are currently VHDL projects for multiple FPGA boards, including
a Gamecube-to-DVI version for the KNJN Pluto IIx-HDMI and a
Gamecube-to-Component/RGB version for a board called GCVideo Lite
which is also documented in this repository.
GCVideo Lite has also been adapted to be used as
an RGB DAC for the Nintendo 64.

The schematics and layout are in the [Hardware](Hardware) directory
and the HDL projects are in the [HDL](HDL) directory. Each directory
should contain README.md files with further information.


## Mini-FAQ ##

1. I want one, how much is it?  
    I do not sell any hardware. Since this is an open-source project,
    other people will probably offer ready-made boards or modding
    services.

    GCVideo-DVI is based on a readily-available, commercial FPGA
    development board, so you could just buy that, flash it and
    install it yourself (or find someone to do it for you).

1. But why don't you just sell it?  
    Building hardware to be sold is a lot of work and requires much
    time that I'd rather use for something more
    interesting. Furthermore, the local laws require quite a bit of
    paperwork and investment to legally sell electronic devices that
    you build and I'd prefer not to deal with all of that.

1. You said your prototype came from OSHPark, so you have at least
    three boards. Can't you sell me one of them?  
    No, they're all accounted for already.

1. Why did you use that weird Video-DAC, it's a non-stock part at
    Digikey!<br>
    It wasn't when I designed the board some months ago... Instead it
    was the cheapest 24-bit video DAC available. It shouldn't be hard
    to adapt the design to a different DAC as long as it has 24 bit
    parallel input and a single-pumped clock. For YPbPr output, the
    DAC also needs to be able to generate a small offset on the Y line
    so that syncs can be generated below the blanking level.

1. Uh, it says XO2-256 in the schematic, but XO2-640 in the BOM. Which
    is the correct one?  
    The original plan was to use an XO2-256, but adding the color
    space conversion for RGB output increased the size of the design
    too much. It was left as an XO2-256 in the schematic to ensure
    only pins available on that chip are used, so a slightly cheaper
    Component-only version can be built.

1. What about line-doubling?  
    GCVideo Lite (the analog version) uses the smallest FPGA that
    could fit everything needed, but this chip does not have enough
    BlockRAM available to generate a line-doubled picture. It could be
    updated by using a larger, footprint-compatible FPGA, but since
    480i/576i are well-supported on component inputs and 240p/288p are
    rarely used by Gamecube titles this has not been attempted yet.

    GCVideo DVI fully supports line-doubling and can also overlay
    scanlines on the line-doubled picture if desired.



# Licence #

<pre>
Copyright (C) 2014-2021, Ingo Korb &lt;ingo@akana.de&gt;
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice,
   this list of conditions and the following disclaimer.
2. Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
THE POSSIBILITY OF SUCH DAMAGE.
</pre>


NEWS
2021-10-16: GCVideo DVI version 3.1
- minor audio subsystem rework
- improve enhanced mode compatibility with some Samsung TVs
- fix incorrect video mode codes in 16:9 enhanced mode
- avoid excessive screen blanking during menu transitions in some games
- add 486p to 480p crop for GBI

2020-05-21: GCVideo DVI version 3.0e
- support another WiiDual board version
- fix clock-related black screen on Wii
- add menu entry for IR key config
- package Wii updater with XML file for Homebrew Channel

2020-02-02: GCVideo DVI version 3.0d
- add GCPlug hardware configuration
- add diagnostics mode in flasher
- fix handling of uncompressed chunks in flasher
- add workaround to never use uncompressed chunks in buildupdate.pl

2020-01-25: GCVideo DVI version 3.0c
- fix stuck display disable on mode change
- attempt to fix a rare audio glitch

2020-01-19: GCVideo DVI version 3.0b
- fix inverted colors on WiiDual
- fix flasher->main reboot failure on some Wiis

2020-01-13: GCVideo DVI version 3.0a
- fix flash corruption when saving settings

2020-01-12: GCVideo DVI version 3.0
- attempted to improve signal compatibility
- non-standard mode detection
- automatic processing bypass for non-standard modes
- automatic image centering, with optional image shift
- more flexible scanline strength including hybrid
- scanline profiles
- infrastructure for installing future firmware updates via console
- faster startup time
- analog component output signal range workaround
- YCbCr 4:4:4 and 4:2:2 output options
- advanced options menu for disabling parts of the image processing
- less strict gamepad snooper timing to fix GC+ compatibility

2019-05-31: GCVideo-DVI version 2.4d
- fix incorrect chroma mixing for Cr-first lines

2018-11-20: GCVideo-DVI version 2.4c
- fix audio scaling

2018-09-14: GCVideo-DVI version 2.4b
- fix a slight chroma shift in YCbCr-to-RGB conversion

2017-12-10: GCVideo-DVI version 2.4a
- fix audio issues introduced by 2.4 for some board configurations
- ensure that the mode popup is always shown with correct colors

2017-10-03: GCVideo-DVI version 2.4
- brightness/contrast/saturation controls in OSD
- add support for dual video output (DVI+parallel RGB, etc. for a DAC)

2016-09-20: GCVideo-DVI version 2.3
- enable IR remote in shuriken/shuriken-v3 builds
- fix a bug that caused OSD CPU hangs on startup

2016-08-14: GCVideo-DVI version 2.2
- Wii version
- add IR remote support to the OSD for p2xh builds

2015-11-22: GCVideo-DVI version 2.1
- audio volume option
- Sync timing fix for enhanced DVI mode

2015-08-21: GCVideo-DVI version 2.0
- Enhanced DVI mode
- optional output disable on mode switch

2015-06-06: GCVideo-DVI version 1.3 and GCVideo Lite [needs-a-version-number]
- Improved SPDIF compatibility

2015-05-17: GCVideo-DVI version 1.2:
- Modify pad snoop timing to support certain NES-to-Gamecube adapters

2015-04-04: GCVideo-DVI version 1.1
- Add OSD and SPDIF audio output

FIXME: Add older entries based on git log


Firmware/pad.h
/* GCVideo DVI Firmware

   Copyright (C) 2015-2021, Ingo Korb <ingo@akana.de>
   All rights reserved.

   Redistribution and use in source and binary forms, with or without
   modification, are permitted provided that the following conditions are met:

   1. Redistributions of source code must retain the above copyright notice,
      this list of conditions and the following disclaimer.
   2. Redistributions in binary form must reproduce the above copyright notice,
      this list of conditions and the following disclaimer in the documentation
      and/or other materials provided with the distribution.

   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
   AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
   IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
   ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
   LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
   CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
   SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
   INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
   CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
   ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
   THE POSSIBILITY OF SUCH DAMAGE.


   pad.h: Gamepad reader and button event generator

*/

#ifndef PAD_H
#define PAD_H

#include <stdint.h>
#include "portdefs.h"
#include "vsync.h"

// 17
#define PAD_LEFT  (1 << 0)
#define PAD_RIGHT (1 << 1)
#define PAD_DOWN  (1 << 2)
#define PAD_UP    (1 << 3)
#define PAD_Z     (1 << 4)
#define PAD_R     (1 << 5)
#define PAD_L     (1 << 6)
// 7 is unused
#define PAD_A     (1 << 8)
#define PAD_B     (1 << 9)
#define PAD_X     (1 << 10)
#define PAD_Y     (1 << 11)
#define PAD_START (1 << 12)
// 29

// mask for all GC buttons
#define PAD_ALL_GC ((1 << 13)-1)

// buttons for short and long hardware "IR-Config" button presses
#define IRBUTTON_SHORT (1 << 13)
#define IRBUTTON_LONG  (1 << 14)

// IR remote buttons
#define IR_UP          (1 << 15)
#define IR_DOWN        (1 << 16)
#define IR_LEFT        (1 << 17)
#define IR_RIGHT       (1 << 18)
#define IR_OK          (1 << 19)
#define IR_BACK        (1 << 20)

// mask for all non-synthetic buttions
#define PAD_ALL        ((1 << 21)-1)

// synthetic button, triggered once on every input video mode change
#define PAD_VIDEOCHANGE (1 << 21)

extern volatile uint32_t pad_buttons;
extern volatile tick_t   pad_last_change;

/* these two functions shouldn't be used in an interrupt handler! */
void pad_set(uint32_t which);
void pad_clear(uint32_t which);

/* same functions without disabling IRQ for atomicity */
void pad_set_irq(uint32_t which);
void pad_clear_irq(uint32_t which);

void pad_wait_for_release(void);
void pad_handler(void);

#endif
